File: CSharp/Applications/AvaloniaSerilogDI/App.axaml.cs
using Avalonia;
using Avalonia.Controls.ApplicationLifetimes;
using Avalonia.Data.Core;
using Avalonia.Data.Core.Plugins;
using Avalonia.Markup.Xaml;
using AvaloniaSerilogDI.ViewModels;
using AvaloniaSerilogDI.Views;
using Common.Core.Extensions;
using LogViewer.Avalonia;
using LogViewer.Core;
using MessageBox.Avalonia;
using MessageBox.Avalonia.Enums;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Logging;
using RandomLogging.Service;
using Serilog;
using Serilog.Sinks.LogView.Core;
using System;
using System.Drawing;
using System.Reflection;
using System.Threading;
using Icon = MessageBox.Avalonia.Enums.Icon;

namespace AvaloniaSerilogDI;

public partial class App : Application
{
    #region Constructors
    
    public override void Initialize()
        => AvaloniaXamlLoader.Load(this); 

    #endregion

    #region Fields

    private IHost? _host;
    private CancellationTokenSource? _cancellationTokenSource;

    #endregion

    #region Methods

     public override void OnFrameworkInitializationCompleted()
    {
        if (ApplicationLifetime is IClassicDesktopStyleApplicationLifetime desktop)
        {
            // Line below is needed to remove Avalonia data validation.
            // Without this line you will get duplicate validations from both Avalonia and CT
            ExpressionObserver.DataValidators.RemoveAll(x => x is DataAnnotationsValidationPlugin);

            // catch all unhandled errors
            AppDomain.CurrentDomain.UnhandledException += OnUnhandledException;

            HostApplicationBuilder builder = Host.CreateApplicationBuilder();

        builder
            // Register the Random Logging Service
            .AddRandomBackgroundService()

            // visual debugging tools
            .AddLogViewer();

        IServiceCollection services = builder.Services;

        // Serilog Logger

        // Azure: https://devblogs.microsoft.com/dotnet/asp-net-core-logging/
        // ApplicationInsights: https://github.com/serilog-contrib/serilog-sinks-applicationinsights
        // AmazonCloudWatch: https://blog.ivankahl.com/logging-dotnet-to-aws-cloudwatch-using-serilog/
        // video: https://www.youtube.com/watch?v=nVAkSBpsuTk (How Structured Logging With Serilog Can Make Your Life Easier)
        // video: https://www.youtube.com/watch?v=_iryZxv8Rxw (C# Logging with Serilog and Seq - Structured Logging Made Easy)
        // ps: docker run -d --restart unless-stopped --name seq -e ACCEPT_EULA=Y -v c:\WIP\LogData:/data -p 8081:80 datalust/seq:latest
        //     docker rmi datalust/seq --force

        // ref: https://stackoverflow.com/questions/66304596/how-to-dependency-inject-serilog-into-the-rest-of-my-classes-in-net-console-app
        services.AddLogging(configure: cfg =>
        {
            Log.Logger = new LoggerConfiguration()
            //Serilog.Core.Logger logger = new LoggerConfiguration()
                .ReadFrom.Configuration(builder.Configuration)
                .WriteTo.DataStoreLoggerSink(
                    dataStoreProvider: () => _host!.Services.TryGetService<ILogDataStore>()!

                    //dataStoreProvider: () => _host!.Services.TryGetService<ILogDataStore>()!,
                    //options =>
                    //{
                    //    options.Colors[LogLevel.Trace] = new()
                    //    {
                    //        Foreground = Color.White,
                    //        Background = Color.DarkGray
                    //    };

                    //    options.Colors[LogLevel.Debug] = new()
                    //    {
                    //        Foreground = Color.White,
                    //        Background = Color.Gray
                    //    };

                    //    options.Colors[LogLevel.Information] = new()
                    //    {
                    //        Foreground = Color.White,
                    //        Background = Color.DodgerBlue
                    //    };

                    //    options.Colors[LogLevel.Warning] = new()
                    //    {
                    //        Foreground = Color.White,
                    //        Background = Color.Orchid
                    //    };
                    //}
                    )
                .CreateLogger();

            cfg.ClearProviders()
                .AddSerilog(Log.Logger);
        });

        services
            .AddSingleton<MainViewModel>()
            .AddSingleton(service => new MainWindow
            {
                DataContext = service.GetRequiredService<MainViewModel>()
            });

            _host = builder.Build();
            _cancellationTokenSource = new();
            
            try
            {
                LogStartingMode();

                // set and show
                desktop.MainWindow = _host.Services.GetRequiredService<MainWindow>();
                desktop.ShutdownRequested += OnShutdownRequested;

                // startup background services
                _ = _host.StartAsync(_cancellationTokenSource.Token);
            }
            catch (OperationCanceledException)
            {
                // skip
            }
            catch (Exception ex)
            {
                Log.Fatal(ex, "Application terminated unexpectedly");

                ShowMessageBox("Unhandled Error", ex.Message);

                CleanUp();
                return;
            }
        }

        base.OnFrameworkInitializationCompleted();
    }

   private void OnShutdownRequested(object? sender, ShutdownRequestedEventArgs e)
        => CleanUp();

    private void OnUnhandledException(object sender, UnhandledExceptionEventArgs e)
    {
        Exception exception = (Exception)e.ExceptionObject;

        Log.Fatal(exception, "Application terminated unexpectedly");
        ShowMessageBox("Unhandled Error", exception.Message);

        CleanUp();
    }

    private void ShowMessageBox(string title, string message)
    {
        MessageBox.Avalonia.BaseWindows.Base.IMsBoxWindow<ButtonResult> messageBoxStandardWindow = MessageBoxManager
            .GetMessageBoxStandardWindow(title, message, ButtonEnum.Ok, Icon.Stop);

        messageBoxStandardWindow.Show();
    }

    private void LogStartingMode()
    {
        // Get the Launch mode
        bool isDevelopment = string.Equals(Environment.GetEnvironmentVariable("DOTNET_MODIFIABLE_ASSEMBLIES"), "debug",
            StringComparison.InvariantCultureIgnoreCase);

        // initialize a logger & EventId
        ILogger<App> logger = _host!.Services.GetRequiredService<ILogger<App>>();
        EventId eventId = new EventId(id: 0, name: Assembly.GetEntryAssembly()!.GetName().Name);

        // log a test pattern for each log level
        logger.TestPattern(eventId: eventId);

        // log that we have started...
        logger.Emit(eventId, LogLevel.Information, $"Running in {(isDevelopment ? "Debug" : "Release")} mode");
    }

    private void CleanUp()
    {
        // tell the background services that we are shutting down
        _ = _host!.StopAsync(_cancellationTokenSource!.Token);

        // flush logs
        Log.CloseAndFlush();
    }

    #endregion
}

File: VB/Applications/AvaloniaNlogVBDI/App.axaml.vb
Imports System.Drawing
Imports System.Reflection
Imports System.Threading
Imports Avalonia
Imports Avalonia.Controls.ApplicationLifetimes
Imports Avalonia.Data.Core
Imports Avalonia.Data.Core.Plugins
Imports Avalonia.Markup.Xaml
Imports LogViewerVB.Avalonia
Imports LogViewerVB.Core
Imports MessageBox.Avalonia
Imports MessageBox.Avalonia.Enums
Imports Microsoft.Extensions.DependencyInjection
Imports Microsoft.Extensions.Hosting
Imports Microsoft.Extensions.Logging
Imports NLog.Target.LogViewVB.Core
Imports RandomLoggingVB.Service

Partial Public Class App
    Inherits Application

    Public Overrides Sub Initialize()
        AvaloniaXamlLoader.Load(Me)
    End Sub

    Public Overrides Sub OnFrameworkInitializationCompleted()

        Dim desktop As IClassicDesktopStyleApplicationLifetime =
                TryCast(ApplicationLifetime, IClassicDesktopStyleApplicationLifetime)

        If desktop IsNot Nothing Then

            ' Line below is needed to remove Avalonia data validation.
            ' Without this line you will get duplicate validations from both Avalonia and CT
            ExpressionObserver.DataValidators.RemoveAll(Function(x) TypeOf x Is DataAnnotationsValidationPlugin)

            ' catch all unhandled errors
            AddHandler AppDomain.CurrentDomain.UnhandledException, AddressOf OnUnhandledException

            Dim builder As HostApplicationBuilder = Host.CreateApplicationBuilder()

            ' 
            ' Note: For information on launch profiles for debugging,
            '     see article: https :  //www.codeproject.com/Articles/5354478/NET-App-Settings-Demystified-Csharp-VB
            ' 

            ' Random Logging Service
            builder.AddRandomBackgroundService()

            ' visual debugging tools
            builder.AddLogViewer()

            'builder.Logging.AddNLogTargets(builder.Configuration)

            ' uncomment to use custom logging colors (note: System.Drawing namespace)
            '
            builder.Logging.AddNLogTargets(
                builder.Configuration,
                Sub(options)

                    options.Colors(LogLevel.Trace) = New LogEntryColor() With
                    {
                        .Foreground = Color.White,
                        .Background = Color.DarkGray
                    }

                    options.Colors(LogLevel.Debug) = New LogEntryColor() With
                    {
                        .Foreground = Color.White,
                        .Background = Color.Gray
                    }

                    options.Colors(LogLevel.Information) = New LogEntryColor() With
                    {
                        .Foreground = Color.White,
                        .Background = Color.DodgerBlue
                    }

                    options.Colors(LogLevel.Warning) = New LogEntryColor() With
                    {
                        .Foreground = Color.White,
                        .Background = Color.Orchid
                    }

                End Sub)

            Dim services As IServiceCollection = builder.Services

            services _
            .AddSingleton(Of MainViewModel) _
            .AddSingleton(Of MainWindow)(
                Function(service) New MainWindow() With
                {
                    .DataContext = service.GetService(Of MainViewModel)
                })

            _host = builder.Build()
            _cancellationTokenSource = New CancellationTokenSource()

            Try
                LogStartingMode()

                ' set and show
                desktop.MainWindow = _host.Services.GetRequiredService(Of MainWindow)()

                ' startup background services
                Dim task As Task = _host.StartAsync(_cancellationTokenSource.Token)

            Catch oce As OperationCanceledException

                ' skip

            Catch ex As Exception

                ShowMessageBox(ex.Message, "Unhandled Error")

            End Try

        End If

        MyBase.OnFrameworkInitializationCompleted()

    End Sub

#Region "Fields"

    Private _host As IHost
    Private _cancellationTokenSource As CancellationTokenSource

#End Region

#Region "Methods"

    Private Sub OnUnhandledException(sender As Object, e As UnhandledExceptionEventArgs)

        ShowMessageBox(DirectCast(e.ExceptionObject, Exception).Message, "Unhandled Error")

    End Sub

    Private Sub ShowMessageBox(title As String, message As String)

        Dim messageBoxStandardWindow = MessageBoxManager _
                .GetMessageBoxStandardWindow(title, message, ButtonEnum.Ok, Enums.Icon.Stop)

        messageBoxStandardWindow.Show()

    End Sub

    Private Sub LogStartingMode()

        ' Get the Launch mode
        Dim isDevelopment As Boolean = String.Equals(Environment.GetEnvironmentVariable("DOTNET_MODIFIABLE_ASSEMBLIES"), "debug",
                                                     StringComparison.InvariantCultureIgnoreCase)


        ' initialize a logger & EventId
        Dim logger As ILogger(Of Application) = _host.Services.GetRequiredService(Of ILogger(Of Application))
        Dim eventId As EventId = New EventId(id:=0, name:=Assembly.GetEntryAssembly.GetName.Name)

        ' log a test pattern for each log level
        logger.TestPattern(eventId)

        ' log that we have started...
        logger.Emit(eventId, LogLevel.Information, $"Running in {If(isDevelopment, "Debug", "Release")} mode")

    End Sub

#End Region

End Class